bftengine/src/bftengine/BFTEngine.cpp

```diff
@@ -1,101 +1,155 @@
 #include "ReplicaImp.hpp"
+#include "DebugPersistentStorage.hpp"
 
-namespace bftEngine
-{
-	namespace impl
-	{
-		struct ReplicaInternal : public Replica
-		{
-			virtual ~ReplicaInternal();
+#include <mutex>
 
-			virtual bool isRunning() const override;
+namespace bftEngine {
+namespace impl {
 
-			uint64_t getLastExecutedSequenceNum() const override;
+bool cryptoInitialized = false;
+std::mutex mutexForCryptoInitialization;
 
-			virtual void start() override;
+struct ReplicaInternal : public Replica {
+  virtual ~ReplicaInternal() override;
 
-			virtual void stop() override;
+  virtual bool isRunning() const override;
 
-                        virtual void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) override;
+  uint64_t getLastExecutedSequenceNum() const override;
 
-			ReplicaImp* rep;
-		};
+  virtual bool requestsExecutionWasInterrupted() const override;
 
+  virtual void start() override;
 
-		ReplicaInternal::~ReplicaInternal()
-		{
-			delete rep;
-		}
+  virtual void stop() override;
 
-		bool ReplicaInternal::isRunning() const
-		{
-			return rep->isRunning();
-		}
+  virtual void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) override;
 
-		uint64_t ReplicaInternal::getLastExecutedSequenceNum() const
-		{
-			return static_cast<uint64_t>(rep->getLastExecutedSequenceNum());
-		}
+  virtual void restartForDebug() override;
 
+  ReplicaImp *rep;
+};
 
-		void ReplicaInternal::start()
-		{
-			return rep->start();
-		}
+ReplicaInternal::~ReplicaInternal() {
+  delete rep;
+}
 
+bool ReplicaInternal::isRunning() const {
+  return rep->isRunning();
+}
 
-		void ReplicaInternal::stop()
-		{
-			return rep->stop();
-		}
+uint64_t ReplicaInternal::getLastExecutedSequenceNum() const {
+  return static_cast<uint64_t>(rep->getLastExecutedSequenceNum());
+}
 
-                void ReplicaInternal::SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a)
-                {
-                        return rep->SetAggregator(a);
-                }
+bool ReplicaInternal::requestsExecutionWasInterrupted() const {
+  const bool run = rep->isRunning();
+  const bool isRecovering = rep->isRecoveringFromExecutionOfRequests();
+  return (!run && isRecovering);
+}
 
-	}
+void ReplicaInternal::start() {
+  return rep->start();
 }
 
+void ReplicaInternal::stop() {
+  return rep->stop();
+}
 
+void ReplicaInternal::SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) {
+  return rep->SetAggregator(a);
+}
 
-bool cryptoInitialized = false;
+void ReplicaInternal::restartForDebug() {
+  Assert(debugPersistentStorageEnabled);
+  rep->stopWhenStateIsNotCollected();
+
+  PersistentStorage *persistentStorage = rep->getPersistentStorage();
+  RequestsHandler *requestsHandler = rep->getRequestsHandler();
+  IStateTransfer *stateTransfer = rep->getStateTransfer();
+  ICommunication *comm = rep->getCommunication();
+
+  // delete rep; TODO(GG): enable after debugging and update ~ReplicaImp
+  rep = nullptr;
```

Comment:
Right, this is a memory leak. This should be fixed together with a change in ~ReplicaImp, which deletes parameters been reused during a test. We can probably use shared pointers here in order to avoid leakage. This code runs only while debugging (debugPersistentStorageEnabled = true and it should be false by default). In any case, we have to consider where to use smart pointers in our code.